### PR TITLE
Fix: Follow discussion settings in the comments block edit

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
This Pull Request aims to respect all the discussion settings in the new Comments block while working in the Site Editor. We are loading some placeholders, and they should follow the discussion settings as much as possible. We decided to show 3 comments (to not overload it with too many comments), but in some cases it wasn't working properly. 

>Bear in mind that, while using the Comments block in the Post Editor, the use case is different. Inside the Site Editor we are using placeholders, but in the Post Editor it gets real data from the REST API, which is a bit trickier. Right now, we are only fetching the top-level comments (due to [this argument](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/comment-template/hooks.js#L29)) and the _embedded children property only includes the next-level children. So we aren’t fetching more than 2 levels at this moment. If we remove the parent: 0  argument, we would get all the comments on the post, but it gets tricky as well. We would have to add some logic to nest the comments and there could be a case where the children are fetched and not the parent, which could cause weird UX issues. Imagine a case where the per_page is set to 10 and we have a parent comment with 15 replies. Only the latest 10 replies would be fetched and the parent would be excluded.
>
> This part is not covered in this Pull Request anyway. 

## Why?
Depending on the discussion settings, the UX right now is a bit weird, so I suggest a new way of handling it to prevent this. For example, at this moment, if the _"Enable threaded (nested) comments X levels deep"_ setting is disabled, just one comment is shown ([this code](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/comment-template/edit.js#L59-L63)).

Screenshot 2022-09-26 at 12 58 40

## How?
I refactored the code a bit and changed slightly the UX. With this I aimed to provide the following workflows:
1. Don't show only two comments if the break comments into pages is disabled (even if it is set to 2):
Screenshot 2022-09-26 at 14 18 06

For this, we have to read the `pageComments` setting.

2. Show at least three top-level comments unless `per_page` is less than three and it is enabled.
Screenshot 2022-09-26 at 14 19 52

3. Show all the nested levels in the first comment if the setting is enabled.
Screenshot 2022-09-26 at 14 20 48

In this case, I wasn't sure if we should limit it to three levels. The max for this option is 10, and I thought it could make sense to show the user how it would look like if the levels is too high.

## Testing Instructions

1. Open the Site Editor with a block-based theme using the new Comments block (the Twenty Twenty-Three theme for example).
2. Navigate to the Single template, where the Comments should be placed.
3. Check the placeholders with the different Discussion Settings.

### CODE REVIEW

**Comment by ockham** [comment]
Thanks a lot for working on this @SantosGuillamot!

I've tried this with my "threaded comments" settings unchanged (which previously got me 3 nested placeholder comments in the Site Editor).

This is now up to 7. IMO, that's too much from a UX point-of-view (and it takes up too much space); after all, they're just placeholders to exemplify a concept 😅 (Also, the heading still states "3 comments" -- it should probably reflect the real number.)

| Before | After |
|--------|------|
| image | image |

**Comment by SantosGuillamot** [comment]
> This is now up to 7. IMO, that's too much from a UX point-of-view.

Yeah, I thought about that too but I wasn't sure. I have just limited it to two nested comments. Anyway, I believe it still makes sense to keep three top-level comments to show the full experience. After all, it is how the posts will look like.

Screenshot 2022-09-27 at 09 01 22


> the heading still states "3 comments" -- it should probably reflect the real number.

You're right, although I'm not sure how this should be handled. The first thing that comes to my mind is to read [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/comments-title/edit.js#L43-L46) the `perPage`, `pageComments`, `threadComments` and `threadCommentsDepth` settings as we are doing in [the Comment Template placeholder](https://github.com/WordPress/gutenberg/blob/fix/discussion-settings-in-comments-block-edit/packages/block-library/src/comment-template/edit.js#L47-L52). And add similar logic there. But I'm not sure it is the best approach. Any ideas?

EDIT: I have just pushed a commit to show what I mean. It works, but we have to keep it synced with the logic changes made to the Comment Template placeholder.

**Comment by ockham** [comment]
Thanks Mario, this is much better!

There's a few minor tweaks I'd still like us to make:

1. TBH, 5 comments are still too much IMO; they take up too much space without providing that much information to users. Let's keep it to 3 or 4.
2. This also applies to the level of nesting: Let's always cut off after 3 levels of nested comments. (I think it's enough to illustrate the idea to users. Just because they have threading set to a large number like 8 doesn't mean we need to show all eight levels in the placeholder.)

TBH, I think we can just leave it at 3. This allows us to hard-wire the Comments Title placeholder back to "3", and gives us the following configurations if I'm not mistaken:

Threaded (`>=3`):

1. Comment
  i. Comment 
    * Comment

Threaded (`==2`):

1. Comment
  i. Comment
3. Comment

Unthreaded:

1. Comment 
2. Comment
3. Comment

I know it doesn't reflect each and every possibility, but I don't think that has to be the goal for a placeholder.

Curious to hear your thoughts!

**Comment by SantosGuillamot** [comment]
Thanks for the feedback 🙂 

Right now, the maximum number of comments is 5, including nested comments. I created a limit at 3 nested comments so no more levels should show. But I was still adding 2 top-level comments if the pagination setting allows it.

Anyways, if it feels there are still too many I can remove the other top-level comments as you suggest to always show 3 comments at max.

> This allows us to hard-wire the Comments Title placeholder back to "3"

Not sure about that. Imagine you have the pagination setting set to 2 comments per page and no nested comments, it would have 2 replies for example.

To be honest, I don't have a strong opinion in this regard, so we can go your way.

**Comment by ockham** [comment]
> Thanks for the feedback 🙂
> 
> Right now, the maximum number of comments is 5, including nested comments. I created a limit at 3 nested comments so no more levels should show. But I was still adding 2 top-level comments if the pagination setting allows it.

Yeah, I thought it was to illustrate that there might be more top-level comments. I think one such top-level comment would be enough to illustrate though, and at the end of the day, I don't even think that it's strictly needed at all.

> Anyways, if it feels there are still too many I can remove the other top-level comments as you suggest to always show 3 comments at max.

Yeah, TBH, I'd rather save more space on screen 😅 

> > This allows us to hard-wire the Comments Title placeholder back to "3"
> 
> Not sure about that. Imagine you have the pagination setting set to 2 comments per page and no nested comments, it would have 2 replies for example.

Not sure I'm following 🤔 Doesn't the Comments Title block show the total number of comments (across all pages)?

Or are you saying it would be confusing for the user if the Comments Title says "3 Comments" but we only show two because of pagination settings? 🤔 That's a fair point.

I'd go as far as to say that only 2 comments per page seems like a bit of an edge case (I'd assume most people would want to show more). However, it might make sense to adapt the Comments Title block logic to reflect the number of comments the user sees in that case 👍 

> To be honest, I don't have a strong opinion in this regard, so we can go your way.

Thanks for bearing with me! 😊

**Comment by SantosGuillamot** [comment]
Okay, I've just pushed a commit to match the UX you defined 🙂 I believe both the comments and the title are working with all the settings. Let me know if there's something not working as expected or missing.

**Comment by ockham** [comment]
tl;dr for triage: This PR is for editor parity with the changes made to the frontend of the Comments block in #44351, to reflect the discussion setting for comments threading (i.e. render comments unthreaded when the setting is disabled). With threading enabled, we'd always show three nested placeholder comments in the Site Editor; with threading disabled, we'd only show one (but still say "3 responses" in the Comments Title block placeholder).

With threading disabled:

image

In the editor:

| `trunk` | This branch |
|--------|------|
| image | image |

---

With threading limited to two levels:

image


| `trunk` | This branch |
|--------|------|
| image | image |

---

With threading limited to three or more levels:

image

| `trunk` | This branch |
|--------|------|
| image | image |

File: packages/block-library/src/comment-template/edit.js [hunk 1]
```diff
@@ -37,6 +37,7 @@ const TEMPLATE = [
  *
  * @param {Object}  settings                       Discussion Settings.
  * @param {number}  [settings.perPage]             - Comments per page setting or block attribute.
+ * @param {boolean}  [settings.pageComments]       - Enable break comments into pages setting.
```
Comment (ockham) [inline packages/block-library/src/comment-template/edit.js]:
```suggestion
 * @param {boolean} [settings.pageComments]        - Enable break comments into pages setting.
```

File: packages/block-library/src/comment-template/edit.js [hunk 2]
```diff
@@ -114,12 +126,12 @@ function CommentTemplateInnerBlocks( {
...
-			is ALWAYS rendered and the preview for the active block is hidden.
-			This ensures that when switching the active block, the component is not
-			mounted again but rather it only toggles the `isHidden` prop.
-
-			The same strategy is used for preventing the flicker in the Post Template
-			block. */ }
+			 is ALWAYS rendered and the preview for the active block is hidden.
+			 This ensures that when switching the active block, the component is not
+			 mounted again but rather it only toggles the `isHidden` prop.
+
```
Comment (ockham) [inline packages/block-library/src/comment-template/edit.js]:
```suggestion

```

**Review by ockham — APPROVED** [review]
Thank you @SantosGuillamot! LGTM :shipit:

**Comment by ockham** [comment]
I just cherry-picked this PR to the wp/6.1 branch to get it included in the next release: b5c56fb60f
